server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSWebOptionsCheckUtility.java

```diff
@@ -214,296 +194,318 @@ public <E extends ParserRuleContext> void checkOptions(E ctx) {
+
+            checkHasIllegalOptions(ctx.SESSTOKEN(), "SESSTOKEN");
+            checkHasIllegalOptions(ctx.CHARACTERSET(), "CHARACTERSET");
         }
 
-        checkPrerequisiteIsMet(ctx.FORMFIELD(), ctx.CHARACTERSET(), ctx, "FORMFIELD");
+        checkPrerequisiteIsMet(ctx.FORMFIELD(), ctx.CHARACTERSET(), ctx, "CHARACTERSET without FORMFIELD");
     }
 
-    private void checkReadnextFormfieldQueryparm(CICSParser.Cics_web_readnext_formfield_queryparmContext ctx) {
-        checkMutuallyExclusiveOptions("FORMFIELD or QUERYPARM", ctx.FORMFIELD(), ctx.QUERYPARM());
+    private void checkReadNext(CICSParser.Cics_web_readnextContext ctx) {
+        checkHasMandatoryOptions(ctx.READNEXT(), ctx, "READNEXT");
 
-        checkHasMandatoryOptions(ctx.NAMELENGTH(), ctx, "NAMELENGTH");
-        checkHasMandatoryOptions(ctx.VALUE(), ctx, "VALUE");
-        checkHasMandatoryOptions(ctx.VALUELENGTH(), ctx, "VALUELENGTH");
-    }
+        checkHasExactlyOneOption("FORMFIELD, HTTPHEADER or QUERYPARM", ctx, ctx.FORMFIELD(), ctx.HTTPHEADER(), ctx.QUERYPARM());
+
+        if (!ctx.FORMFIELD().isEmpty() || !ctx.QUERYPARM().isEmpty()) {
+            checkHasIllegalOptions(ctx.SESSTOKEN(), "SESSTOKEN");
+        }
 
-    private void checkReadnextHTTPHeader(CICSParser.Cics_web_readnext_httpheaderContext ctx) {
         checkHasMandatoryOptions(ctx.NAMELENGTH(), ctx, "NAMELENGTH");
         checkHasMandatoryOptions(ctx.VALUE(), ctx, "VALUE");
         checkHasMandatoryOptions(ctx.VALUELENGTH(), ctx, "VALUELENGTH");
     }
 
-    private void checkReceiveServerBuffer(CICSParser.Cics_web_receive_server_bufferContext ctx) {
-        checkHasMandatoryOptions(ctx.LENGTH(), ctx, "LENGTH");
-    }
+    private void checkReceive(CICSParser.Cics_web_receiveContext ctx) {
+        checkHasMandatoryOptions(ctx.RECEIVE(), ctx, "RECEIVE");
 
-    private void checkReceiveServerContainer(CICSParser.Cics_web_receive_server_containerContext ctx) {
-        checkHasMandatoryOptions(ctx.TOCONTAINER(), ctx, "TOCONTAINER");
-    }
+        if (!ctx.INTO().isEmpty() || !ctx.SET().isEmpty() || !ctx.LENGTH().isEmpty() || !ctx.MAXLENGTH().isEmpty() || !ctx.NOTRUNCATE().isEmpty() || !ctx.SRVCONVERT().isEmpty() || !ctx.NOSRVCONVERT().isEmpty() || !ctx.SERVERCONV().isEmpty()
+                || !ctx.CLICONVERT().isEmpty() || !ctx.NOCLICONVERT().isEmpty() || !ctx.CLIENTCONV().isEmpty() || !ctx.HOSTCODEPAGE().isEmpty()) {
+            // Buffer
+            checkHasIllegalOptions(ctx.TOCONTAINER(), "TOCONTAINER");
+            checkHasIllegalOptions(ctx.TOCHANNEL(), "TOCHANNEL");
 
-    private void checkReceiveClient(CICSParser.Cics_web_receive_clientContext ctx) {
-        checkHasMandatoryOptions(ctx.SESSTOKEN(), ctx, "SESSTOKEN");
-    }
+            checkHasExactlyOneOption("INTO or SET", ctx, ctx.INTO(), ctx.SET());
+            checkMutuallyExclusiveOptions("SRVCONVERT, NOSRVCONVERT or SERVERCONV", ctx.SRVCONVERT(), ctx.NOSRVCONVERT(), ctx.SERVERCONV());
```

Comment:
CLICONVERT needs to be checked as well. Also check will depend on the client vs server variant.

Maybe reorganize this a bit to have first 4 bools - buffer/container x client/server, then check the common stuff and then 4(-6) if blocks for the specific parts?